engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoMultiValue.java

```diff
@@ -83,285 +99,264 @@ public final String toDisplayString(boolean ignore) {
-    for (var i = 0; i < values.length; i++) {
-      if (iop.fitsInByte(values[i])) {
-        return iop.asByte(values[i]);
-      }
+    var both = InteropType.find(values, methodDispatchTypes, iop);
+    if (both.type() == InteropType.NUMBER) {
+      return iop.asByte(both.value());
     }
     throw UnsupportedMessageException.create();
   }
 
   @ExportMessage
   short asShort(@Shared("interop") @CachedLibrary(limit = "10") InteropLibrary iop)
       throws UnsupportedMessageException {
-    for (var i = 0; i < values.length; i++) {
-      if (iop.fitsInShort(values[i])) {
-        return iop.asShort(values[i]);
-      }
+    var both = InteropType.find(values, methodDispatchTypes, iop);
+    if (both.type() == InteropType.NUMBER) {
+      return iop.asShort(both.value());
     }
     throw UnsupportedMessageException.create();
   }
 
   @ExportMessage
   int asInt(@Shared("interop") @CachedLibrary(limit = "10") InteropLibrary iop)
       throws UnsupportedMessageException {
-    for (var i = 0; i < values.length; i++) {
-      if (iop.fitsInInt(values[i])) {
-        return iop.asInt(values[i]);
-      }
+    var both = InteropType.find(values, methodDispatchTypes, iop);
+    if (both.type() == InteropType.NUMBER) {
+      return iop.asInt(both.value());
     }
     throw UnsupportedMessageException.create();
   }
 
   @ExportMessage
   long asLong(@Shared("interop") @CachedLibrary(limit = "10") InteropLibrary iop)
       throws UnsupportedMessageException {
-    for (var i = 0; i < values.length; i++) {
-      if (iop.fitsInLong(values[i])) {
-        return iop.asLong(values[i]);
-      }
+    var both = InteropType.find(values, methodDispatchTypes, iop);
+    if (both.type() == InteropType.NUMBER) {
+      return iop.asLong(both.value());
     }
     throw UnsupportedMessageException.create();
   }
 
   @ExportMessage
   float asFloat(@Shared("interop") @CachedLibrary(limit = "10") InteropLibrary iop)
       throws UnsupportedMessageException {
-    for (var i = 0; i < values.length; i++) {
-      if (iop.fitsInFloat(values[i])) {
-        return iop.asFloat(values[i]);
-      }
+    var both = InteropType.find(values, methodDispatchTypes, iop);
+    if (both.type() == InteropType.NUMBER) {
+      return iop.asFloat(both.value());
     }
     throw UnsupportedMessageException.create();
   }
 
   @ExportMessage
   double asDouble(@Shared("interop") @CachedLibrary(limit = "10") InteropLibrary iop)
       throws UnsupportedMessageException {
-    for (var i = 0; i < values.length; i++) {
-      if (iop.fitsInDouble(values[i])) {
-        return iop.asDouble(values[i]);
-      }
+    var both = InteropType.find(values, methodDispatchTypes, iop);
+    if (both.type() == InteropType.NUMBER) {
+      return iop.asDouble(both.value());
     }
     throw UnsupportedMessageException.create();
   }
 
   @ExportMessage
   boolean fitsInBigInteger(@Shared("interop") @CachedLibrary(limit = "10") InteropLibrary iop) {
```

Comment:
nit: You don't need to fill in the "interop" name inside `@Shared` annotation, as long as the parameter name (in this case `iop`) is the same for all the methods with `@Shared` parameter. Feel free to keep it explicitly there.